Code-4-Community/constellation
# Set up QLDB and TypeScript

### ℹ️ Issue

Closes #5 

### 📝 Description

- Add TypeScript support to the backend — this involves changing our build script from `sam build` to `npx tsc && sam build` which is equivalent to the newly created script `npm run build`
- Set up QLDB
  - Create a ledger
  - Create an AWS KMS key for encryption/decryption of data in QLDB
  - Create an IAM role to give the Lambdas permission to interact with QLDB
- Add sample code for creating tables, creating indexes, inserting documents, and querying documents.

### ✔️ Verification

What steps did you take to verify your changes work? These should be clear enough for someone to be able to clone the branch and follow the steps themselves. 

1. `npm run build`
2. `sam deploy --capabilities CAPABILITY_NAMED_IAM`
3. Try the three endpoints: `curl <URL>`, curl<URL>/1`, `curl -d '{"ids":"1", "name":"meg"}' -H "Content-Type: application/json" -X POST <URL>`
4. Visit AWS CloudWatch and inspect the logs from the Lambdas.
5. Visit AWS QLDB and query the table.

### 🏕️ (Optional) Future Work / Notes

Two options for table creation:
1. Before running any database transaction in the code, check if the table exists and if not, create the table.
2. Assume the table has already been created in the application code and make sure to manually create the table before deploying.

I prefer option 2; it simplifies our code and prevents a lot of database transaction calls that check if the table exists.
